build/gen_stub.php

```diff
@@ -1339,82 +1492,517 @@ public function getMethodSynopsisElement(array $funcMap, array $aliasMap, DOMDoc
+            return "    REGISTER_LONG_CONSTANT(\"$constName\", " . ($cConstValue ?: (int) $constValue) . ", $flags);\n";
+        }
+
+        if ($value->type->isFloat()) {
+            return "    REGISTER_DOUBLE_CONSTANT(\"$constName\", " . ($cConstValue ?: (float) $constValue) . ", $flags);\n";
+        }
+
+        if ($value->type->isString()) {
+            return "    REGISTER_STRING_CONSTANT(\"$constName\", " . ($cConstValue ?: '"' . addslashes($constValue) . '"') . ", $flags);\n";
+        }
+
+        throw new Exception("Unimplemented constant type");}
+
+    /**
+     * @param iterable<ConstInfo> $allConstInfos
+     */
+    private function getClassConstDeclaration(EvaluatedValue $value, iterable $allConstInfos): string
+    {
+        $constName = $this->getVariableLikeName();
+
+        $zvalCode = $value->initializeZval("const_{$constName}_value", $allConstInfos);
+
+        $code = "\n" . $zvalCode;
+
+        $code .= "\tzend_string *const_{$constName}_name = zend_string_init_interned(\"$constName\", sizeof(\"$constName\") - 1, 1);\n";
+        $nameCode = "const_{$constName}_name";
+
+        $code .= "\tzend_declare_class_constant_ex(class_entry, $nameCode, &const_{$constName}_value, " . $this->getFlagsAsString() . ", NULL);\n";
+        $code .= "\tzend_string_release(const_{$constName}_name);\n";
+
+        return $code;
     }
 
-    public function __clone()
+    private function getValueAssertion(EvaluatedValue $value): string
     {
-        foreach ($this->args as $key => $argInfo) {
-            $this->args[$key] = clone $argInfo;
+        if ($value->isUnknownConstValue || $value->originatingConst || $value->cConstName === null) {
+            return "";
         }
-        $this->return = clone $this->return;
-    }
-}
 
-function initializeZval(string $zvalName, $value): string
-{
-    $code = "\tzval $zvalName;\n";
+        $cName = $value->cConstName;
+        $constValue = $value->value;
 
-    switch (gettype($value)) {
-        case "NULL":
-            $code .= "\tZVAL_NULL(&$zvalName);\n";
-            break;
+        if ($value->type->isNull()) {
+            return "    ZEND_ASSERT($cName == NULL);\n";
+        }
 
-        case "boolean":
-            $code .= "\tZVAL_BOOL(&$zvalName, " . ((int) $value) . ");\n";
-            break;
+        if ($value->type->isBool()) {
+            $cValue = $constValue ? "true" : "false";
+            return "    ZEND_ASSERT($cName == $cValue);\n";
+        }
 
-        case "integer":
-            $code .= "\tZVAL_LONG(&$zvalName, $value);\n";
-            break;
+        if ($value->type->isInt()) {
+            $cValue = (int) $constValue;
+            return "    ZEND_ASSERT($cName == $cValue);\n";
+        }
 
-        case "double":
-            $code .= "\tZVAL_DOUBLE(&$zvalName, $value);\n";
-            break;
+        if ($value->type->isFloat()) {
+            $cValue = (float) $constValue;
+            return "    ZEND_ASSERT($cName == $cValue);\n";
```

Comment:
I'm wondering whether this suffers from precision issues? I suppose at the top of the script we need to use `ini_set("precision", 17);`

Reply:
Hmm, yes, you must be right!